File: scroll.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main(int argc, const char* argv[]) {
  if (argc < 2) {
    printf("usage: %s <string> [offset] [max-length]\n", argv[0]);
    return 1;
  }

  unsigned long length = strlen(argv[1]);

  if (length == 0) {
    return 0;
  }

  unsigned long size = length;
  unsigned long offset = 0;

  if (argc > 2) {
    offset = atoi(argv[2]);
  }

  if (argc > 3) {
    size = atoi(argv[3]);
  }

  if (size <= 0) {
    return 0;
  } else if (length < size) {
    printf("%s", argv[1]);
    return 0;
  }

  char output[size + 1];
  unsigned long index = 0;
  while (index < size) {
    output[index] = argv[1][(index + offset) % length];
    index++;
  }

  output[size] = '\0';

  printf("%s", output);
  return 0;
}

File: scroll.lua
local substr = function (text, begin_index, end_index)
  local codepoints = vim.str_utf_pos(text)

  local last_byte_size = vim.str_utf_end(text, codepoints[end_index])
  return string.sub(
    text,
    codepoints[begin_index],
    codepoints[end_index] + last_byte_size
  )
end
local strlen = function (text)
  return #vim.str_utf_pos(text)
end
local scroll = function (text, offset, length)
  local text_length = strlen(text)
  if text_length <= length then
    return text
  end
  text = text .. '   '
  text_length = strlen(text)

  local adjusted_offset = offset % text_length
  local output = substr(
    text,
    adjusted_offset + 1,
    math.min(adjusted_offset + length, text_length)
  )

  local remaining = length - strlen(output)
  while remaining > 0 do
    if text_length <= remaining then
      output = output .. text
    else
      output = output .. substr(
        text,
        1,
        remaining
      )
    end
    remaining = length - strlen(output)
  end

  return output
end

if #arg < 3 then
  -- print('usage: lua scroll.lua <text> <offset> <length>')
  os.exit(1)
end
print(scroll(arg[1], tonumber(arg[2]), tonumber(arg[3])))
